File: www/highlight.ts
export type TokenKind =
  | 'keyword'
  | 'string'
  | 'number'
  | 'comment'
  | 'function'
  | 'operator'
  | 'punctuation'
  | 'property'
  | 'variable'
  | 'text';

export interface Token {
  kind: TokenKind;
  value: string;
}

const KEYWORDS = new Set([
  'workflow', 'fn', 'var', 'if', 'else', 'foreach', 'in', 'on', 'return',
  'true', 'false', 'null', 'import', 'from',
]);

const BUILTIN_FUNCS = new Set([
  'log', 'len', 'to_string', 'to_number',
]);

export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;

  while (i < input.length) {
    const ch = input[i];

    // Comments
    if (ch === '/' && input[i + 1] === '/') {
      let end = input.indexOf('\n', i);
      if (end === -1) end = input.length;
      tokens.push({ kind: 'comment', value: input.slice(i, end) });
      i = end;
      continue;
    }

    // Strings
    if (ch === '"') {
      let j = i + 1;
      while (j < input.length && input[j] !== '"') {
        if (input[j] === '\\') j++;
        j++;
      }
      tokens.push({ kind: 'string', value: input.slice(i, j + 1) });
      i = j + 1;
      continue;
    }

    // Numbers
    if (/[0-9]/.test(ch) || (ch === '-' && i + 1 < input.length && /[0-9]/.test(input[i + 1]))) {
      let j = i;
      if (input[j] === '-') j++;
      while (j < input.length && /[0-9]/.test(input[j])) j++;
      if (j < input.length && input[j] === '.') {
        j++;
        while (j < input.length && /[0-9]/.test(input[j])) j++;
      }
      tokens.push({ kind: 'number', value: input.slice(i, j) });
      i = j;
      continue;
    }

    // Identifiers and keywords
    if (/[a-zA-Z_]/.test(ch)) {
      let j = i;
      while (j < input.length && /[a-zA-Z0-9_]/.test(input[j])) j++;
      const word = input.slice(i, j);

      if (KEYWORDS.has(word)) {
        tokens.push({ kind: 'keyword', value: word });
      } else if (BUILTIN_FUNCS.has(word)) {
        tokens.push({ kind: 'function', value: word });
      } else if (j < input.length && input[j] === '(') {
        tokens.push({ kind: 'function', value: word });
      } else {
        tokens.push({ kind: 'variable', value: word });
      }
      i = j;
      continue;
    }

    // Multi-char operators
    if (i + 1 < input.length) {
      const two = input.slice(i, i + 2);
      if (two === '==' || two === '!=' || two === '<=' || two === '>=' || two === '&&' || two === '||') {
        tokens.push({ kind: 'operator', value: two });
        i += 2;
        continue;
      }
    }

    // Single-char operators
    if ('+-*/%=<>!'.includes(ch)) {
      tokens.push({ kind: 'operator', value: ch });
      i++;
      continue;
    }

    // Punctuation
    if ('(){}[],.'.includes(ch)) {
      tokens.push({ kind: 'punctuation', value: ch });
      i++;
      continue;
    }

    // Whitespace and other
    tokens.push({ kind: 'text', value: ch });
    i++;
  }

  return tokens;
}

File: crates/workflow-native-editor/src/highlight.rs
use eframe::egui::Color32;
use std::collections::HashSet;

#[derive(Clone, Copy)]
pub enum TokenKind {
    Keyword,
    String,
    Number,
    Comment,
    Function,
    Operator,
    Punctuation,
    Variable,
}

pub struct Token {
    pub text: String,
    pub kind: TokenKind,
}

/// Tokenize a line of code with awareness of known function names.
/// The `known_functions` parameter allows the tokenizer to recognize
/// functions registered in the dynamic FunctionRegistry.
pub fn tokenize_line(line: &str, known_functions: &HashSet<String>) -> Vec<Token> {
    let mut tokens = Vec::new();
    let mut i = 0;
    let bytes = line.as_bytes();

    while i < bytes.len() {
        let ch = bytes[i] as char;

        if ch == '/' && i + 1 < bytes.len() && bytes[i + 1] as char == '/' {
            tokens.push(Token {
                text: line[i..].to_string(),
                kind: TokenKind::Comment,
            });
            break;
        }

        if ch == '"' {
            let start = i;
            i += 1;
            while i < bytes.len() {
                if bytes[i] as char == '"' {
                    i += 1;
                    break;
                }
                if bytes[i] as char == '\\' {
                    i += 2;
                } else {
                    i += 1;
                }
            }
            tokens.push(Token {
                text: line[start..i].to_string(),
                kind: TokenKind::String,
            });
            continue;
        }

        if ch.is_ascii_digit()
            || (ch == '-' && i + 1 < bytes.len() && (bytes[i + 1] as char).is_ascii_digit())
        {
            let start = i;
            if ch == '-' {
                i += 1;
            }
            while i < bytes.len() && (bytes[i] as char).is_ascii_digit() {
                i += 1;
            }
            if i < bytes.len() && bytes[i] as char == '.' {
                i += 1;
                while i < bytes.len() && (bytes[i] as char).is_ascii_digit() {
                    i += 1;
                }
            }
            tokens.push(Token {
                text: line[start..i].to_string(),
                kind: TokenKind::Number,
            });
            continue;
        }

        if ch.is_ascii_alphabetic() || ch == '_' {
            let start = i;
            i += 1;
            while i < bytes.len()
                && ((bytes[i] as char).is_ascii_alphanumeric() || bytes[i] as char == '_')
            {
                i += 1;
            }
            let word = &line[start..i];
            let kind = match word {
                "workflow" | "fn" | "var" | "if" | "else" | "foreach" | "in" | "on" | "return"
                | "true" | "false" | "null" | "import" | "from" | "emit" => TokenKind::Keyword,
                _ => {
                    // Check if it's a known function (from registry or builtins)
                    if known_functions.contains(word) {
                        TokenKind::Function
                    } else {
                        TokenKind::Variable
                    }
                }
            };
            tokens.push(Token {
                text: word.to_string(),
                kind,
            });
            continue;
        }

        if "+-*/%=<>!&|".contains(ch) {
            let start = i;
            if i + 1 < bytes.len() {
                let two = &line[i..i + 2];
                if matches!(two, "==" | "!=" | "<=" | ">=" | "&&" | "||") {
                    i += 2;
                    tokens.push(Token {
                        text: two.to_string(),
                        kind: TokenKind::Operator,
                    });
                    continue;
                }
            }
            i += 1;
            tokens.push(Token {
                text: line[start..i].to_string(),
                kind: TokenKind::Operator,
            });
            continue;
        }

        if "(){}[],.".contains(ch) {
            tokens.push(Token {
                text: ch.to_string(),
                kind: TokenKind::Punctuation,
            });
            i += 1;
            continue;
        }

        tokens.push(Token {
            text: ch.to_string(),
            kind: TokenKind::Variable,
        });
        i += 1;
    }

    tokens
}

pub fn token_color(kind: TokenKind) -> Color32 {
    match kind {
        TokenKind::Keyword => Color32::from_rgb(200, 120, 255),
        TokenKind::String => Color32::from_rgb(180, 220, 120),
        TokenKind::Number => Color32::from_rgb(255, 180, 100),
        TokenKind::Comment => Color32::from_gray(100),
        TokenKind::Function => Color32::from_rgb(100, 200, 255),
        TokenKind::Operator => Color32::from_rgb(255, 120, 120),
        TokenKind::Punctuation => Color32::from_gray(180),
        TokenKind::Variable => Color32::WHITE,
    }
}
